src/python/pants/scm/subsystems/changed.py

```diff
@@ -160,10 +162,24 @@ def register_options(cls, register):
         )
 
 
+@dataclass(frozen=True)
+class UncachedScmWrapper:
+    """???/the salt is intended to be different every time, so the scm should avoid being memoized
+    by the engine!"""
+
+    scm: Scm
+    salt: str
+
+    @classmethod
+    def create(cls, scm: Scm) -> "UncachedScmWrapper":
+        return cls(scm=scm, salt=str(uuid.uuid4()),)
```

Comment:
I wonder if you could use the new _uncachable_rule mechanism added by Benjy today in https://github.com/pantsbuild/pants/pull/10403?

Reply:
Yes, that PR was based off of this exact work when I mentioned it to benjy and I haven't edited after rebasing yet because inferring per-file targets is now breaking.

Reply:
Done!